post_processing_stages/object_classify_tf_stage.cpp

```diff
@@ -0,0 +1,370 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright (C) 2021, Raspberry Pi (Trading) Limited
+ *
+ * negate_stage.cpp - image negate effect
+ */
+
+#include <algorithm>
+#include <fstream>
+#include <functional>
+#include <iostream>
+#include <map>
+#include <memory>
+#include <queue>
+#include <string>
+#include <vector>
+
+#include <libcamera/stream.h>
+
+#include "core/libcamera_app.hpp"
+#include "core/post_processing_stage.hpp"
+
+#include "tensorflow/lite/builtin_op_data.h"
+#include "tensorflow/lite/c/common.h"
+#include "tensorflow/lite/interpreter.h"
+#include "tensorflow/lite/kernels/register.h"
+#include "tensorflow/lite/string_util.h"
+
+using Stream = libcamera::Stream;
+
+struct TfSettings
+{
+	TfLiteType input_type = kTfLiteFloat32;
+	bool allow_fp16 = false;
+	tflite::string model_name = "/home/pi/models/mobilenet_quant_v1_224.tflite";
+	tflite::FlatBufferModel *model;
+	tflite::string labels_file_name = "/home/pi/models/labels.txt";
+	int number_of_threads = 4;
+	int number_of_results = 3;
+};
+
+constexpr int IMAGE_WIDTH = 224;
+constexpr int IMAGE_HEIGHT = 224;
+constexpr int IMAGE_CHANNELS = 3;
+
+class ObjectClassifyTfStage : public PostProcessingStage
+{
+public:
+	ObjectClassifyTfStage(LibcameraApp *app) : PostProcessingStage(app) {}
+
+	char const *Name() const override;
+
+	void Read(boost::property_tree::ptree const &params);
+
+	void Configure();
+
+	bool Process(CompletedRequest &completed_request);
+
+	void Stop();
+
+private:
+	void initialize();
+
+	void runInference();
+
+	TfLiteStatus readLabelsFile(const tflite::string &file_name, std::vector<tflite::string> *result,
+								size_t *found_label_count);
+
+	std::vector<uint8_t> yuvToRgb(uint8_t *mem, int w, int h, int stride);
+
+	void getTopResults(uint8_t *prediction, int prediction_size, size_t num_results,
+					   std::vector<std::pair<float, int>> *top_results, TfLiteType input_type);
+
+	Stream *stream_;
+	int w_, h_, stride_;
+	std::mutex future_ptr_mutex_;
+	std::mutex output_results_mutex_;
+	TfSettings settings_;
+	std::unique_ptr<tflite::FlatBufferModel> model_;
+	std::unique_ptr<tflite::Interpreter> interpreter_;
+	const std::vector<int> inputs_;
+	const std::vector<int> outputs_;
```

Comment:
I don't think we're using either inputs_ or outputs_? They might want to be deleted.

Reply:
Deleted both of them.